### ISSUE TITLE
PHP7 REWRITE SESSION 2.3.x  and 2.4

### ISSUE THREAD

**Issue by oitsuki** [osCommerce/oscommerce2#387]
In configuration select rewrite session on true
go to catalog, try to connect and make an order.
After the login you will have a session problem with session_regenerate_id(); (function session) and register_shutdown_function('session_write_close');

file sessions.php and all files where there is the call above

If you rewrite session is on false, all work fine.

Warning: session_regenerate_id(): Session callback expects true/false return value
Warning: session_regenerate_id(): Session write failed. ID: user (path: /tmp/)

**Comment by ruden** [comment]
+1

**Comment by Gergely** [comment]
Could you test it in sessions.php?

```php
    function _sess_close() {
      tep_db_close();
      return true;
    }
```

**Comment by oitsuki** [comment]
No doesn't work, 

I insert a comment in this line and also in configuration recreate on false
```
// php7 pb
//   register_shutdown_function('session_write_close');
```

**Comment by Gergely** [comment]
and if you use file session?

my problem is that its working for me

```php
  define('STORE_SESSIONS', 'file');
```

**Comment by oitsuki** [comment]
define('STORE_SESSIONS', 'file');
recreate session on true
decomment
// php7 pb
`   register_shutdown_function('session_write_close');`

seems to be better and work on base 2.4

**Comment by Gergely** [comment]
I have no problem in official standard v2.3.4 but has problem in v2.4 on the same enviroment.
Differents are program loading logic so I think its a PHP7 bug.

**Comment by Gergely** [comment]
@oitsuki 

do you have oscomid? Probably this will be the problem. I dont see sesskey in session table and oscomid in cookies. Its empty.

**Comment by Gergely** [comment]
Workaround the problem the _sess_write() function return value always should be **"true"**
It seems to be valid from PHP5.4 where

```php
register_shutdown_function('session_write_close');
```
**has no effect** or its working slient in the backgound when user function applied. Maybe PDO class protection is the reason.

So your solution where decomented the _register_shutdown_function()_ do the same efffect at this moment.

Official v2.3.4 use _db_query_ functions only and its working perfectly on PHP7.

**Comment by oitsuki** [comment]
Ok, I hope Harald will begin shortly the 2.4 to continue and finish the development

**Comment by oitsuki** [comment]
I think it can come because mysql 7 is in strict mode, some errors appears in this case
for example
the correct synthax will be for request (int), (float => decimal) are important in this case
```
          $OSCOM_PDO->save('products_to_categories', [ 'products_id' => (int)$products_id ,
                                                                                        'categories_id' => (int)$current_category_id
                                                                                        ]
                                               );
```

Just I want to say with the session, it will be maybe the same thing, strict mode can make problem. It could be a raison some bugs can appear and some code must rewrite correctly to work in strict mode.
I will try without the strict mode and we will see.

**Comment by haraldpdl** [comment]
This does not seem to be related to MySQL strict mode but the php session save handle functions not returning their proper values. This has been fixed in v2.4, and for v2.3 with the following commit:

https://github.com/osCommerce/oscommerce2/commit/d2d4d7d05355aa7f04732cfc27495bd07af42998
